### ISSUE TITLE
jax.pswapaxes doesn't work with "non-input" tensors

### ISSUE THREAD

**Issue by Thenerdstation** [google/jax#1651]
Currently, it is impossible to use `jax.lax.pswapaxes` on tensors that don't either come directly from the input or are the result of computation from the inputs.

Examples:

This works fine
```python
def test(a):
  return jax.lax.pswapaxes(a, "i", 0)

jax.pmap(test, "i")(
    np.arange(16 * 16).reshape(16, 16))
```

However, if I were to ignore the `a` input value and instead try to do a swapping on a new tensor created, it fails with a confusing error.

```python
def test(a):
  del a
  a = np.ones(16)
  return jax.lax.pswapaxes(a, "i", 0)

jax.pmap(test, "i")(
    np.arange(16 * 16).reshape(16, 16))
```

This is the resulting error.

```
jax/interpreters/pxla.py in apply_parallel_primitive(prim, *args, **params)
    289   else:
    290     shape = (logical_size(dynamic_axis_env[axis_name]),)
--> 291   return parallel_pure_rules[prim](*args, shape=shape, **params)
    292 
    293 parallel_pure_rules = {}

KeyError: all_to_all
```

**Comment by Thenerdstation** [comment]
One could imagine needing to do this if the tensor you're working with is too large to give as a single input and you need to generate it on the fly.

You can mostly work around this issue by generating it in a different pmap function and then using that as an input, but from a user perspective it might be better if this "just worked" instead.

**Comment by levskaya** [comment]
You're correct that `pswapaxes` doesn't work on a non-data-dependent input, and in general most parallel operators cannot, given the current model of data-dependent tracing used by pmap.

In fact, the only constant that pmap currently handles without staging out is the special case `axis_size = psum(1, 'axis_name')` which is where these "parallel pure" rules come into play.

So if you need to send out giant constants, we recommend you do so using a 
"sharded device put" (equivalent to `pmap(lambda x: x)`)  or a 
"replicated device put" (equivalent to `lambda x: pmap(lambda: x)`) in the case of copying the same thing N times.  

Now, we haven't actually yet committed the memory-efficient version of the latter into JAX yet, but PR #1562 has the code for the replicated device put and we hope to get it in soon.

**Comment by rajasekharporeddy** [comment]
Hi @chaserileyroberts 
I executed the mentioned code with JAX version 0.3.25 on Google Colab TPU. Since the number of XLA devices available are 8 in free version of Colab TPU, I modified the code accordingly and it executed without any error. Looks like `pswapaxes` works for the 'non-input' tensors on later versions of JAX.
```python
import jax
import jax.numpy as jnp
import numpy as np

def test(a):
  del a
  a = np.ones(8)
  return jax.lax.pswapaxes(a, "i", 0)

jax.pmap(test, "i")(
    np.arange(8 * 8).reshape(8, 8))
```
Output:
```
ShardedDeviceArray([[1., 1., 1., 1., 1., 1., 1., 1.],
                    [1., 1., 1., 1., 1., 1., 1., 1.],
                    [1., 1., 1., 1., 1., 1., 1., 1.],
                    [1., 1., 1., 1., 1., 1., 1., 1.],
                    [1., 1., 1., 1., 1., 1., 1., 1.],
                    [1., 1., 1., 1., 1., 1., 1., 1.],
                    [1., 1., 1., 1., 1., 1., 1., 1.],
                    [1., 1., 1., 1., 1., 1., 1., 1.]], dtype=float32)
```
Kindly find the [gist](https://colab.research.google.com/gist/rajasekharporeddy/7c0925ba70c65d13a5312d54b0a80b36/-1651.ipynb) for reference.

Thank you
